File: src/dores/Analisador.java
package dores;

import dores.enums.OperadorRelacional;
import dores.enums.OperadorAritmetico;
import dores.enums.SimboloPontuacao;
import dores.enums.Tipo;
import dores.estruturas.Token;

import java.io.BufferedReader;
import java.io.FileInputStream;
import java.io.InputStreamReader;
import java.io.PushbackReader;
import java.util.ArrayList;
import java.util.List;

public class Analisador {

    private String path;

    private List<Token> tokens = new ArrayList<>();

    private PushbackReader r;

    private int coluna = 0;

    private int linha = 1;

    private int estado = 0;

    public Analisador(String path) {
        this.path = path;
    }

    public List<Token> Analisar() {
        try {
            r = new PushbackReader(new BufferedReader(new InputStreamReader(
                    new FileInputStream(path), "US-ASCII")));

            Character data;
            while ((data = Ler()) != -1) {
                if (Character.isWhitespace((data))) {
                    proximaLinha(data);
                    continue;
                }
                Token token = pegarToken(data);
                if(fimAnalisar(token)){
                    return tokens;
                }
            }
        } catch (Exception e) {
            System.out.println(e);
        }

        return tokens;
    }

    // TODO: LINHA E COLUNA
    private Token pegarToken(Character c) {
        String lexema = "";
        int colunaAtual = 0;
        int linhaAtual = 0;

        while (true) {
            switch (estado) {
                case 0:
                    colunaAtual = coluna;
                    linhaAtual = linha;
                    if (Character.isDigit(c)) {
                        lexema += c;
                        c = Ler();
                        estado = 12;
                    } else if (Character.isAlphabetic(c)) {
                        lexema += c;
                        c = Ler();
                        estado = 14;
                    } else if ( c.equals(':')) {
                        lexema += c;
                        c = Ler();
                        estado = 1;
                    } else if (OperadorRelacional.charToSimbol(c) || OperadorAritmetico.charToSimbol(c) ||
                                SimboloPontuacao.charToSimbol(c) && !c.equals('=')) {
                        variosIfsSemFim(c);
                        lexema += c;
                    } else {
                        return null;
                    }
                    break;
                case 1:
                    if (c.equals('=')) {
                        lexema += c;
                        estado = 2;
                    } else {
                        desler(c);
                        estado = 3;
                    }
                    break;
                case 2:
                    return retornaToken(Tipo.SATRIBUICAO,lexema,linhaAtual,colunaAtual);
                case 3:
                    return retornaToken(Tipo.STIPO,lexema,linhaAtual,colunaAtual);
                case 4:
                    return retornaToken(Tipo.SMAIS,lexema,linhaAtual,colunaAtual);
                case 5:
                    return retornaToken(Tipo.SMENOS,lexema,linhaAtual,colunaAtual);
                case 6:
                    return retornaToken(Tipo.SMULTIPLICACAO,lexema,linhaAtual,colunaAtual);
                case 7:
                    return retornaToken(Tipo.SDIVIDIR,lexema,linhaAtual,colunaAtual);
                case 8:
                    return retornaToken(Tipo.SPONTO_E_VIRGULA,lexema,linhaAtual,colunaAtual);
                case 9:
                    return retornaToken(Tipo.SPONTO,lexema,linhaAtual,colunaAtual);
                case 10:
                    return retornaToken(Tipo.SABRE_PARENTESIS,lexema,linhaAtual,colunaAtual);
                case 11:
                    return retornaToken(Tipo.SFECHA_PARENTESIS,lexema,linhaAtual,colunaAtual);
                case 12:
                    if (Character.isDigit(c)) {
                        lexema += c;
                        c = Ler();
                    } else if (!Character.isDigit(c)) {
                        desler(c);
                        estado = 13;
                    }
                    break;
                case 13:
                    return retornaToken(Tipo.SNUMERO,lexema,linhaAtual,colunaAtual);
                case 14:
                    if (Character.isDigit(c) || Character.isAlphabetic(c) || c.equals('_')) {
                        lexema += c;
                        c = Ler();
                    } else if (!Character.isDigit(c) || !Character.isAlphabetic(c) || !c.equals(')')) {
                        desler(c);
                        estado = 15;
                    }
                    break;
                case 15:
                    return retornaToken(Tipo.SIDENTIFICADOR,lexema,linhaAtual,colunaAtual);
                default:
                   return retornaToken(Tipo.SERRO,lexema,linhaAtual,colunaAtual);
            }
        }
    }

    private Character Ler() {
        try {
            coluna++;
            int c = r.read();
            return c != -1 ? (char) c : (char) -1;
        } catch (Exception e) {
            System.out.println(e);
        }
        return (char) -1;
}

    private void proximaLinha(Character c) {
        if (c.equals('\n')) {
            linha++;
            coluna = 0;
        }
    }

    // gostei do nome, mt criativo
    private void desler(Character ch) {
        try {
            coluna--;
            r.unread((int) ch);
        } catch (Exception e) {
            System.out.println(e);
        }
    }

    private void variosIfsSemFim(Character c) {
        if (c.equals('+')) {
            estado = 4;
        } else if (c.equals('-')) {
            estado = 5;
        } else if (c.equals('*')) {
            estado = 6;
        } else if (c.equals('/')) {
            estado = 7;
        } else if (c.equals(';')) {
            estado = 8;
        } else if (c.equals('.')) {
            estado = 9;
        } else if (c.equals('(')) {
            estado = 10;
        } else if (c.equals(')')) {
            estado = 11;
        }
    }

    private boolean fimAnalisar(Token token){
        if (token == null) {
            return true;
        } else if(Tipo.SERRO.equals(token.getTipo())){
            tokens.add(token);
            return true;
        } else {
            tokens.add(token);
            return false;
        }
    }

    private Token retornaToken(Tipo tipo, String lexema, int linha, int coluna ){
        estado = 0;
        return new Token(tipo,lexema,linha,coluna);
    }
}

File: c#/Compiladoress/Compiladoress/Analisador.cs
using Compiladoress.Estrutura;
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;

namespace Compiladoress
{
    public class Analisador
    {
        private readonly string _path;
        private List<Token> tokens = new List<Token>();
        private int linha = 1;
        private int coluna = 0;
        private int estado = 0;
        private StreamReader r;
        public Analisador(string path)
        {
            _path = path;
        }
        ~Analisador()
        {
            r.Dispose();
            r.Close();
        }

        public List<Token> Analisar()
        {
            try
            {
                r = new StreamReader(this._path, encoding: Encoding.ASCII);
                char c;
                int cI;
                while ((cI = r.Peek()) != -1)
                {
                    c = (char)cI;

                    if (Char.IsWhiteSpace(c))
                    {
                        proximaLinha(c);
                        continue;
                    }
                    c = (char)r.Read(); coluna++;
                    Token t = pegarToken(c);
                    if (t == null) return tokens;
                    tokens.Add(t);
                    if(t.Tipo == EnumTipo.SERRO) return tokens;
                }
            }
            catch (Exception e)
            {
                Console.WriteLine(e.Message);
            }
            return tokens;
        }

        private Token pegarToken(char c)
        {
            string lexema = "";
            int colunaAtual = 0;
            int linhaAtual = 0;
            while (true)
            {
                switch (this.estado)
                {
                    case 0:
                        colunaAtual = coluna;
                        linhaAtual = linha;
                        if (Char.IsDigit(c))
                        {
                            lexema += c;
                            estado = 12;
                        }
                        else if (Char.IsLetter(c))
                        {
                            lexema += c;
                            estado = 14;
                        }
                        else if (':'.Equals(c))
                        {
                            lexema += c;
                            estado = 1;
                        }
                        else if ((estado = OperadorAritmetico.simboloValido(c)) != -1)
                        {
                            lexema += c;
                        }
                        else if ((estado = SimboloPontuacao.simboloValido(c)) != -1)
                        {
                            lexema += c;
                        }
                        else
                        {
                            return null;
                        }
                        break;
                    case 1:
                        if ('='.Equals(c))
                        {
                            c = (char)r.Read(); coluna++;
                            lexema += c;
                            estado = 2;
                        }
                        else
                        {
                            estado = 3;
                        }
                        break;
                    case 2: return ReturnToken(EnumTipo.SATRIBUICAO, linhaAtual, colunaAtual, lexema);
                    case 3: return ReturnToken(EnumTipo.STIPO, linhaAtual, colunaAtual, lexema);
                    case 4: return ReturnToken(EnumTipo.SMAIS, linhaAtual, colunaAtual, lexema);
                    case 5: return ReturnToken(EnumTipo.SMENOS, linhaAtual, colunaAtual, lexema);
                    case 6: return ReturnToken(EnumTipo.SMULTIPLICACAO, linhaAtual, colunaAtual, lexema);
                    case 7: return ReturnToken(EnumTipo.SDIVIDIR, linhaAtual, colunaAtual, lexema);
                    case 8: return ReturnToken(EnumTipo.SPONTO_E_VIRGULA, linhaAtual, colunaAtual, lexema);
                    case 9: return ReturnToken(EnumTipo.SPONTO, linhaAtual, colunaAtual, lexema);
                    case 10: return ReturnToken(EnumTipo.SABRE_PARENTESIS, linhaAtual, colunaAtual, lexema);
                    case 11: return ReturnToken(EnumTipo.SFECHA_PARENTESIS, linhaAtual, colunaAtual, lexema);
                    case 12:
                        if (Char.IsDigit(c))
                        {
                            lexema += c;
                            c = (char)r.Read(); coluna++;
                        }
                        else
                        {
                            estado = 13;
                        }
                        break;
                    case 13:
                        return ReturnToken(EnumTipo.SNUMERO, linhaAtual, colunaAtual, lexema);
                    case 14:
                        if (Char.IsLetterOrDigit(c) || '_'.Equals(c))
                        {
                            lexema += c;
                            c = (char)r.Read(); coluna++;
                        }
                        else if(!Char.IsLetterOrDigit(c) || !')'.Equals(c)) {
                            estado = 15;
                        }
                        break;
                    case 15: return ReturnToken(EnumTipo.SIDENTIFICADOR, linhaAtual, colunaAtual, lexema);
                    default: return ReturnToken(EnumTipo.SERRO, linhaAtual, colunaAtual, lexema);

                }
                c = (char)r.Peek();
            }
        }
        private void proximaLinha(Char c)
        {
            
            if (c.Equals('\n'))
            {
                linha++;
                coluna = 0;
                c = (char)r.Read();
            }
            else
            {
                c = (char)r.Read(); coluna ++;
            }
            
        }
        private Token ReturnToken(EnumTipo tipo, int linha, int coluna, string lexema)
        {
            estado = 0;
            return new Token
            {
                Tipo = tipo,
                Coluna = coluna,
                Linha = linha,
                Lexema = lexema
            };
        }
    }
}
